src/dex/uniswap-v3/uniswap-v3.ts

```diff
@@ -0,0 +1,824 @@
+          ],
+        },
+        poolIdentifier: this.getPoolIdentifier(
+          pool.token0,
+          pool.token1,
+          pool.feeCode,
+        ),
+        exchange: this.dexKey,
+        gasCost: UNISWAPV3_QUOTE_GASLIMIT,
+        poolAddresses: [pool.poolAddress],
+      };
+    });
+
+    return result;
+  }
+
+  async getPricesVolume(
+    srcToken: Token,
+    destToken: Token,
+    amounts: bigint[],
+    side: SwapSide,
+    blockNumber: number,
+    limitPools?: string[],
+  ): Promise<null | ExchangePrices<UniswapV3Data>> {
+    try {
+      const _srcToken = this.dexHelper.config.wrapETH(srcToken);
+      const _destToken = this.dexHelper.config.wrapETH(destToken);
+
+      const [_srcAddress, _destAddress] = this._getLoweredAddresses(
+        _srcToken,
+        _destToken,
+      );
+
+      if (_srcAddress === _destAddress) return null;
+
+      let selectedPools: UniswapV3EventPool[] = [];
+      if (limitPools === undefined) {
+        for (const fee of this.supportedFees) {
+          let pool =
+            this.eventPools[
+              this.getPoolIdentifier(_srcAddress, _destAddress, fee)
+            ];
+          if (!pool) {
+            pool = await this.getPool(
+              _srcAddress,
+              _destAddress,
+              fee,
+              blockNumber,
+            );
+          }
+          if (pool) {
+            selectedPools.push(pool);
+          }
+        }
+      } else {
+        const pairIdentifierWithoutFee = this.getPoolIdentifier(
+          _srcAddress,
+          _destAddress,
+          0n,
+          // Trim from 0 fee postfix, so it become comparable
+        ).slice(0, -1);
+
+        const poolIdentifiers = limitPools.filter(identifier =>
+          identifier.startsWith(pairIdentifierWithoutFee),
+        );
+        for (const identifier of poolIdentifiers) {
+          let pool = this.eventPools[identifier];
+          if (!pool) {
+            const [, srcAddress, destAddress, fee] = identifier.split('_');
+            pool = await this.getPool(
+              srcAddress,
+              destAddress,
+              BigInt(fee),
+              blockNumber,
+            );
+          }
+
+          if (pool) {
+            selectedPools.push(pool);
+          }
+        }
+      }
+
+      if (selectedPools.length === 0) return null;
+
+      const poolsToUse = selectedPools.reduce(
+        (acc, pool) => {
+          let state = pool.getState(blockNumber);
+          if (state === null) {
+            this.logger.trace(
+              `${this.dexKey}: State === null. Fallback to rpc ${pool.name}`,
+            );
+            acc.poolWithoutState.push(pool);
+          } else {
+            acc.poolWithState.push(pool);
+          }
+          return acc;
+        },
+        {
+          poolWithState: [] as UniswapV3EventPool[],
+          poolWithoutState: [] as UniswapV3EventPool[],
+        },
+      );
+
+      const rpcResultsPromise = this.getStateFromRpc(
```

Comment:
If I understood it correctly, we expect that RPC calls should be very low. In that case, we need to add logging, to know how frequently we are using RPC calls.